bpo-42927: Inline cache for slots

gvanrossum opened the pull request:
This is a modification of the `LOAD_ATTR` inline cache that landed a few months ago (PR 22803).

The design overloads the `hint` field: if it is negative, it indicates a slot. Slots are created using `__slots__`, but some builtin types that use `PyMemberDescrObject` descriptors (with the type set to `T_OBJECT_EX`) will also benefit.

I've tried to keep the code style similar to the existing `LOAD_ATTR` cache (happy path first) but I had to make some compromises, e.g. the cache initialization code can no longer skip everything if `tp_dictoffset` is zero (since it will be zero for classes with only slots).

<!-- issue-number: [bpo-42927](https://bugs.python.org/issue42927) -->
https://bugs.python.org/issue42927
<!-- /issue-number -->

gvanrossum committed:
Implement inline cache for slots

1st1 commented on Python/ceval.c:
```diff
@@ -3173,6 +3174,25 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
+                            }
+                            else {
+                                Py_DECREF(owner);
+                                goto error;
```
Perhaps we want to `OPCACHE_DEOPT_LOAD_ATTR();` here.

gvanrossum replied:
But why would we? The slot is still there, it's just empty. The hint is still correct.

1st1 replied:
Ah, right, makes sense.

pablogsal commented on Python/ceval.c:
```diff
@@ -3221,50 +3241,81 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
+                                    // First time we optimize this opcode.
+                                    OPCACHE_STAT_ATTR_OPT();
+                                    co_opcache->optimized = OPCODE_CACHE_MAX_TRIES;
+                                    fprintf(stderr, "Setting hint for %s, offset %zd\n", dmem->name, offset);
```
I think this should be removed :)

gvanrossum replied:
I commented it out, in case my future self wants to play more with this.

pablogsal commented on Python/ceval.c:
```diff
@@ -3173,6 +3174,25 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
+                            }
+                            else {
+                                Py_DECREF(owner);
+                                goto error;
```
This is going to the error label without setting an exception. Is this an error case or should we just go to the regular path ?

1st1 replied:
~We could just raise an AttributeError here, but~ I'd say falling back and going the regular well tested path is a safer option.

pablogsal replied:
It may be easier to just raise than make it so it goes through the right sequence of conditionals without adding more goto labels. Unless I miss something we just need to:

```
                               SET_TOP(NULL);
                               _PyErr_SetString(tstate, PyExc_AttributeError, .........);
                               goto error;
```

1st1 replied:
Wouldn't that break `__getattr__`?

```
>>> class F:
...   __slots__ = ('a',)
...   def __init__(self):
...    self.a = 1
...   def __getattr__(self, k):
...    print(self, k)
...
>>> f = F()
>>> f.a
1
>>> f.b
<__main__.F object at 0x10bbfb070> b
>>>
```

pablogsal replied:
~We have this to populate the chache~:

```
                if (co_opcache != NULL &&  type->tp_getattro == PyObject_GenericGetAttr)
```

~Unless I am seeing another code path, if you add a __getattr__ to a class it will never get optimized~

Sorry I had in mind `__getattribute__` (which is `tp_getattro` not `tp_getattr`).

1st1 replied:
Yeah, I think we should follow the already established way and simply fallback to the slow path (calling `PyObject_GetAttr(owner, name);`) if the slot is unused.

pablogsal replied:
Wait a second, isn't `__getattr__` **also** `tp_getattro`?:

https://github.com/python/cpython/blob/f4507231e3f0cf8827cec5592571ce371c6813e8/Objects/typeobject.c#L7250


I think we should be fine because we are not optimizing in that case. Maybe I am missing something. In any case, I agree that we should try to fall back to the slow path unless is much more complicated.

1st1 replied:
> In any case, I agree that we should try to fall back to the slow path unless is much more complicated.

Yeah, it would be easier to review/read/reason about the code that way.

gvanrossum replied:
Sorry I didn't see this before. We should do exactly what happens in `PyMember_GetOne()` for `case T_OBJECT_EX:`. It does this:
```
PyErr_SetString(PyExc_AttributeError, l->name);
```
So I'll add that.

gvanrossum replied:
Oh wait, the name isn't available in the hint. So I'll just fall back to the fast path as Pablo said.

pablogsal commented on Python/ceval.c:
```diff
@@ -3221,50 +3241,81 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
                                 }
+                                else {
+                                    Py_DECREF(owner);
+                                    goto error;
```
Same thing here as https://github.com/python/cpython/pull/24216/files#r557650392

gvanrossum committed:
Comment out debug print

gvanrossum committed:
Fall through to slow path if slot is NULL

blurb-it[bot] committed:
📜🤖 Added by blurb_it.

gvanrossum committed:
Fix the buildbot test failure (I hope).

The problem was that sometimes a hint of -1 could be set, meaning the
dict lookup failed.  My code mistook that for a slot offset.

The fix is simple, because slot offsets can never be 0, so ~offset
can never be -1.

I also cleaned up some comments and a variable name.

gvanrossum committed:
Merge remote-tracking branch 'origin/master' into slotscache

1st1 commented on Python/ceval.c:
```diff
@@ -3174,9 +3174,10 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
-                        // Hint >= 0 is a dict index; hint < 0 is an inverted slot index.
-                        if (la->hint < 0) {
-                            /* Even faster path -- slot hint */
+                        // Hint >= 0 is a dict index; hint == -1 is a dict miss.
```
Oh, right! How did I miss that on review :/

gvanrossum replied:
Same way I missed it writing the code. :-(

This code is hairy.

1st1 commented on Python/ceval.c:
```diff
@@ -3174,9 +3174,10 @@ _PyEval_EvalFrameDefault(PyThreadState *tstate, PyFrameObject *f, int throwflag)
-                        if (la->hint < 0) {
-                            /* Even faster path -- slot hint */
+                        // Hint >= 0 is a dict index; hint == -1 is a dict miss.
+                        // Hint < -1 is an inverted slot offset (offsets are never 0).
```
Maybe clarify that slot offsets are always greater than 1?

gvanrossum replied:
I have a comment ready to push:
```
                        // Hint >= 0 is a dict index; hint == -1 is a dict miss.
                        // Hint < -1 is an inverted slot offset: offset is strictly > 0,
                        // so ~offset is strictly < -1 (assuming 2's complement).
```
Let me know if you still want improvement.

1st1 replied:
looks good to me

gvanrossum committed:
Improve comment
